### PR TITLE
Use weekly tables in the chunk store

### PR DESCRIPTION
**Pull request by tomwilkie**
Part of #158

Introduces a new job, the table manager, which periodically calculates the required tables (and provisioned throughputs) and then creates / updates the existing tables as appropriate.

Also update the chunk store to calculate the tables a given bucket should be written to.

TODO:
- [x] update `bigBuckets` to also pick a table for the bucket
- [x] new binary/container to create / update tables
- [x] unit test
- [x] add some monitoring for the dynamodb operations.
- [x] figure out when we can correctly remove write capacity from a table (see #182)

### CODE REVIEW

File: chunk/chunk_store.go [hunk 1]
```diff
@@ -475,15 +407,15 @@ func (c *AWSStore) updateIndex(ctx context.Context, userID string, chunks []Chun
...
 	}
 
-	return c.batchWriteDynamo(ctx, c.tableName, writeReqs)
+	return c.batchWriteDynamo(ctx, writeReqs)
 }
 
 // calculateDynamoWrites creates a set of batched WriteRequests to dynamo for all
 // the chunks it is given.
 //
 // Creates one WriteRequest per bucket per metric per chunk.
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: not necessarily - most of the time, buckets will like in the same table.

File: chunk/chunk_store.go [hunk 2]
```diff
@@ -475,15 +407,15 @@ func (c *AWSStore) updateIndex(ctx context.Context, userID string, chunks []Chun
...
-func (c *AWSStore) calculateDynamoWrites(userID string, chunks []Chunk) ([]*dynamodb.WriteRequest, error) {
-	writeReqs := []*dynamodb.WriteRequest{}
+func (c *AWSStore) calculateDynamoWrites(userID string, chunks []Chunk) (map[string][]*dynamodb.WriteRequest, error) {
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: I thought about just plumbing the whole `bucketSpec` through...  WDYT?

File: chunk/chunk_store.go [hunk 3]
```diff
@@ -980,9 +924,12 @@ func (r *dynamoBatchWriteItemsOp) do() {
 		}
 
 		// If there are unprocessed items, backoff and retry those items.
-		if resp.UnprocessedItems != nil && len(resp.UnprocessedItems[r.tableName]) > 0 {
-			dynamoUnprocessedItems.Add(float64(len(resp.UnprocessedItems[r.tableName])))
-			unprocessed = append(unprocessed, resp.UnprocessedItems[r.tableName]...)
+		if resp.UnprocessedItems != nil && dictLen(resp.UnprocessedItems) > 0 {
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: The reality is most of the time we'll only be writing to a single table, and in the worse case we'll be writing to two (when we are on the edge of two tables).  This is nice a clear so I think I'd prefer to leave it as is.

File: chunk/chunk_store.go [hunk 4]
```diff
@@ -143,61 +142,24 @@ type Store interface {
...
+	ChunkCache  *Cache
+
+	// After midnight on this day, we start bucketing indexes by day instead of by
+	// hour.  Only the day matters, not the time within the day.
 	DailyBucketsFrom model.Time
 
+	UsePeriodicTables    bool
+	TablePrefix          string
+	TablePeriod          time.Duration
+	PeriodicTableStartAt time.Time
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: Correct.  👍

File: chunk/chunk_store.go [hunk 5]
```diff
@@ -349,32 +268,45 @@ func (c *AWSStore) CreateTables() error {
 // This function deals with any changes from one bucketing scheme to another -
 // for instance, it know the date at which to migrate from hourly buckets to
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: 👍

File: chunk/chunk_store.go [hunk 6]
```diff
@@ -349,32 +268,45 @@ func (c *AWSStore) CreateTables() error {
...
-		result []string
+		result []bucketSpec
 	)
 
+	tableForBucket := func(bucketStart int64) string {
+		if !c.cfg.UsePeriodicTables || bucketStart < (c.cfg.PeriodicTableStartAt.Unix()) {
+			return c.tableName
+		}
+		return c.cfg.TablePrefix + strconv.Itoa(int(bucketStart/int64(c.cfg.TablePeriod/time.Second)))
+	}
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: 👍

File: chunk/chunk_store.go [hunk 7]
```diff
@@ -942,31 +872,45 @@ func (r *dynamoQueryPagesOp) do() {
...
-	fillReq := func(in *[]*dynamodb.WriteRequest, out *[]*dynamodb.WriteRequest) {
-		outLen, inLen := len(*out), len(*in)
+	dictLen := func(in map[string][]*dynamodb.WriteRequest) int {
+		result := 0
+		for _, reqs := range in {
+			result += len(reqs)
+		}
+		return result
+	}
+	fillReq := func(in map[string][]*dynamodb.WriteRequest, out map[string][]*dynamodb.WriteRequest) {
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: 👍

File: chunk/chunk_store.go [hunk 8]
```diff
@@ -991,7 +938,11 @@ func (r *dynamoBatchWriteItemsOp) do() {
 		// If we get provisionedThroughputExceededException, then no items were processed,
 		// so back off and retry all.
 		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == provisionedThroughputExceededException {
-			unprocessed = append(unprocessed, reqs...)
+			for tableName, unprocessReqs := range reqs {
+				unprocessed[tableName] = append(unprocessed[tableName], unprocessReqs...)
+				dynamoUnprocessedItems.Add(float64(len(unprocessReqs)))
+			}
```
Comment (tomwilkie) [inline chunk/chunk_store.go]: 👍

File: chunk/chunk_store_test.go [hunk 9]
```diff
@@ -61,16 +71,26 @@ func TestChunkStoreUnprocessed(t *testing.T) {
...
 
+	tableManager, err := NewDynamoTableManager(TableManagerConfig{
+		dynamodb: dynamodb,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := tableManager.syncTables(context.Background()); err != nil {
+		t.Fatal(err)
+	}
```
Comment (tomwilkie) [inline chunk/chunk_store_test.go]: 👍

File: chunk/dynamo_table_manager.go [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+
+// TableManagerConfig is the config for a DynamoTableManager
+type TableManagerConfig struct {
+	DynamoDBURL          string
+	DynamoDBPollInterval time.Duration
+
+	UsePeriodicTables    bool
+	TablePrefix          string
+	TablePeriod          time.Duration
+	PeriodicTableStartAt time.Time
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: 👍

File: chunk/dynamo_table_manager.go [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+
+	// duration a table will be created before it is needed.
+	CreationGracePeriod        time.Duration
+	MaxChunkAge                time.Duration
+	ProvisionedWriteThroughput int64
+	ProvisionedReadThroughput  int64
+
+	// Not exported as only used by tests to inject mocks
+	dynamodb dynamodbClient
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: Considered this, but I'm trying to move stuff out of main.go

File: chunk/dynamo_table_manager.go [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+		dynamodb:  dynamodbClient,
+		tableName: tableName,
+
+		cfg:  cfg,
+		done: make(chan struct{}),
+	}
+	return m, nil
+}
+
+// Start starts the DynamoTableManager
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: 👍

File: chunk/dynamo_table_manager.go [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+func (m *DynamoTableManager) Start() {
+	m.wait.Add(1)
+	go m.loop()
+}
+
+// Stop stops the DynamoTableManager
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: 👍

File: chunk/dynamo_table_manager.go [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+	toCreate, toCheckThroughput, err := m.partitionTables(ctx, expected)
+	if err != nil {
+		return err
+	}
+
+	if err := m.createTables(ctx, toCreate); err != nil {
+		return err
+	}
+
+	return m.updateTables(ctx, toCheckThroughput)
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: No reason, I think its easiest just to fail everything - we'll need monitoring / alerting on this either way.

File: chunk/dynamo_table_manager.go [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+	provisionedWrite int64
+}
+
+type byName []tableDescription
+
+func (a byName) Len() int           { return len(a) }
+func (a byName) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byName) Less(i, j int) bool { return a[i].name < a[j].name }
+
+func (m *DynamoTableManager) calculateExpectedTables(_ context.Context) []tableDescription {
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: consistency.

File: chunk/dynamo_table_manager.go [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+		return m.dynamodb.ListTablesPages(&dynamodb.ListTablesInput{}, func(resp *dynamodb.ListTablesOutput, _ bool) bool {
+			for _, s := range resp.TableNames {
+				existingTables = append(existingTables, *s)
+			}
+			return true
+		})
+	}); err != nil {
+		return nil, nil, err
+	}
+	sort.Strings(existingTables)
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: 👍

File: chunk/dynamo_table_manager.go [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+	i, j := 0, 0
+	for i < len(descriptions) && j < len(existingTables) {
+		if descriptions[i].name < existingTables[j] {
+			// Table descriptions[i] doesn't exist
+			toCreate = append(toCreate, descriptions[i])
+		} else if descriptions[i].name > existingTables[j] {
+			// existingTables[j].name isn't in descriptions, can ignore
+			j++
+		} else {
+			// Table existis, need to check it has correct throughput
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: 👍

File: chunk/dynamo_table_manager.go [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+			toCheckThroughput = append(toCheckThroughput, descriptions[i])
+			i++
+			j++
+		}
+	}
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: :pike:!!!

File: chunk/dynamo_table_manager.go [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+	for ; i < len(descriptions); i++ {
+		toCreate = append(toCreate, descriptions[i])
+	}
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: Now we've broken out the listTables function, this is basically all thats left.  Still want me to do it?

File: chunk/chunk_store.go (hunk 2 above)
Comment (tomwilkie) [inline chunk/chunk_store.go]: done.

File: chunk/dynamo_table_manager.go [hunk 20]
```diff
@@ -0,0 +1,347 @@
...
+		// if we are before the switch to periodic table, we need to give this table write throughput
+		if now < (firstTable*tablePeriodSecs)+gracePeriodSecs+maxChunkAgeSecs {
+			legacyTable.provisionedWrite = m.cfg.ProvisionedWriteThroughput
+		}
+		result = append(result, legacyTable)
+	}
+
+	for i := firstTable; i <= lastTable; i++ {
+		table := tableDescription{
+			// Name construction needs to be consistent with chunk_store.bigBuckets
```
Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: What would that interface look like?  Note here we need something different to in chunk_store: we want all tables from the first to now.  In chunk store we want the table for a given time.

Comment (jml) [inline chunk/dynamo_table_manager.go]: Not 100% sure, and getting sure would probably require doing the refactoring myself.

File: chunk/dynamo_table_manager.go (hunk 19 above)
Comment (jml) [inline chunk/dynamo_table_manager.go]: I do (it's easier to test & re-use that way!) but I can understand why you don't want to.

File: chunk/dynamo_table_manager.go (hunk 14 above)
Comment (jml) [inline chunk/dynamo_table_manager.go]: Agreed re monitoring. It's just that the error handling convenience here kind of implies a logical flow when there isn't one. /shrug

File: chunk/chunk_store.go (hunk 2 above)
Comment (jml) [inline chunk/chunk_store.go]: I like this better, thanks.

File: chunk/chunk_store.go (hunk 1 above)
Comment (jml) [inline chunk/chunk_store.go]: ah I see, it's "grouped by table".

File: chunk/dynamo_table_manager.go (hunk 11 above)
Comment (jml) [inline chunk/dynamo_table_manager.go]: The `MakeDynamoDbClient` function can still be defined here. That'd keep main.go fairly small.

Comment (tomwilkie) [inline chunk/dynamo_table_manager.go]: Done.

**Comment by juliusv** [comment]
Given that @jml has done a deeper review, I've just given this a 30-minute read to get an impression of how it all works. It looks overall good to me. The complexity that both the code and the system are reaching now scares me, but I don't have a simpler alternative to offer either. So a rough :+1: - and crazy that you can produce so much code that fast :)

File: cmd/cortex_table_manager/main.go [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+		},
+	}
+
+	addr := flag.String("web.listen-addr", ":80", "The address to listen on for HTTP requests.")
+	flag.DurationVar(&cfg.DynamoDBPollInterval, "dynamodb.poll-interval", 2*time.Minute, "How frequently to poll DynamoDB to learn our capacity.")
+	periodicTableStartAt := flag.String("dynamodb.periodic-table.start", "", "DynamoDB periodic tables start time.")
+	dynamodbURL := flag.String("dynamodb.url", "localhost:8000", "DynamoDB endpoint URL.")
+	flag.StringVar(&cfg.TablePrefix, "dynamodb.periodic-table.prefix", "cortex_", "DynamoDB table prefix for the periodic tables.")
+	flag.DurationVar(&cfg.TablePeriod, "dynamodb.periodic-table.period", 7*24*time.Hour, "DynamoDB periodic tables period.")
+	flag.DurationVar(&cfg.CreationGracePeriod, "dynamodb.periodic-table.grace-period", 10*time.Minute, "DynamoDB periodic tables grace period (duration which table will be created/delete before/after its needed).")
```
Comment (juliusv) [inline cmd/cortex_table_manager/main.go]: its -> it's

Comment (juliusv) [inline cmd/cortex_table_manager/main.go]: delete -> deleted

**Pull request merged**